### ISSUE TITLE
Unable to read very large files because of use of binary bit shifts.

### ISSUE THREAD

**Issue by manticorp** [photopea/UTIF.js#19]
Take the line:

```
var bytes = new Uint8Array((img.width*img.height*bipp) >> 3);
```

For a very large image, this will cause a RangeError.

Take for example a 16535 x 9652 with bipp = 24, the result should be 478787460, but instead in javascript we get -58083452

By changing all instances of ```>>3``` to ```/8``` and all instances of ```<<3``` to ```*8``` we can open much larger files than previously.

So, the previous example would look something like this:

```
var bytes = new Uint8Array((img.width*img.height*bipp)/8);
```

However, when there are a weird number of bits per pixel, these numbers aren't necessarily integers, so I guess we'd need to do some floor/ceiling on the numbers?

Anyway, this solved an issue I had opening huge TIFF files using this library - **was wondering if anyone had any thoughts?**

***

To expand, javascript binary operations are 32 bits max:

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Bitwise_Operators

So any file with > 2^31 pixels will automatically not be openable, just because of these bitshifts.

**Comment by photopea** [comment]
That is an interesting bug. I guess you are the firt person to open so large images with UTIF.js

What if you keep the previous version, but replace all >> with >>> ? Does it help?

**Comment by manticorp** [comment]
This seems to have solved it. I just replaced all instances of ```>>``` with ```>>>```, and it worked.

I can now open and decompress very large tiffs.

Now to find a way of putting those humungous images on a canvas! Every time I try, it seems to crash my browser.

Thanks for the help!

**Comment by photopea** [comment]
I did the same thing in UTIF.js here at GitHub, so please, update it to keep being synchronized.

Your image has around 0.5 GB after the decompression. Allocating so huge canvas, creating ImageData for it, would take another 1 GB. Why do you even do it? Do you really have a display, that can show 16535 x 9652 pixels?

If not, I would suggest scaling it down. E.g. ten times: convert each 10x10px square of an old image into a 1x1px square of a new image and give it an average color of from those original 100 pixels.

**Comment by manticorp** [comment]
Awesome, thank you.

Hmm...an interesting thought...I will investigate this, and see if I can get a decent resizing algorithm in javascript.

Largest *real* canvas size I expect to experience is roughly ~3200px wide and ~6000px tall (a 3x zoom overlay on the original canvas, so that the zoomed canvas element has a real size of 3x the on screen display and all movements are amplified 3x), and my application should allow for scaling of the original image up to 500%.

I suppose I will have to accept that there will be some compromises. I just thought that there might be some internal optimizations on the browser side that would let me have such a large amount of pixel data.

For reference, this is for a print related application, and people regularly use our current solution with images that are 1.5m wide and 5m long at 300DPI - it all works server side atm, and am trying to find a browser side solution.

I will have to do some experimenting.

**Comment by photopea** [comment]
There is no problem in manipulating your image in the browser. Just create the canvas according to the display screen. Either the user has zoomed out, so you show e.g. the 10x smaller version of the image. Or the user is zoomed in, but then you just copy the corresponding sub-area (e.g 1920x1080px) of your image. And when user scrolls the mouse wheel, you gradually update it with a shifted sub-area of the image.

In other words, you will never need to put the whole image into a canvas, or even have two copies of your huge image.

**Comment by manticorp** [comment]
Hmm - yes, this is true. I think I will have to go down this route.

I've managed to adapt some nice resizing algorithms (hermite, catrom, lanczos, etc) to work with my application now, so I think I will use a cached, resized image for the general display, and 1:1 slices of the data for the zoom.

Just for your interest, opening a 16535 x 9652px LZW compressed tiff file takes 29s (all time is taken in the decoding/decompression stage), and doing a hermite resize of the image from 16535px to 2048px data takes 2.1s - this is on my Core i5, 8GB memory, 2017 Dell XPS 13 laptop (fairly mid range in terms of processing power).

The decoding/decompression step could probably be sped up a fair amount by paralellizing using workers. I will look into this.

**Comment by photopea** [comment]
Maybe I could speed up the LZW decoding by optimizing my code. Could you send me your tiff file (or some other hard LZW file) to support@photopea.com ?

BTW. why do you use TIFF and not PNG?

**Comment by manticorp** [comment]
It's not me who's using TIFFs, it's a customer facing application. People *love* tiffs. 

I've had a look at the decompression code (only looking at LZW for the mo) - I'm currently just trying to understand it at the moment! Without really learning how LZW compression works, it's hard for me to even begin changing things in anything but the most naive manner. My solution is usually to just chuck more resources at it (hence using workers, can be multithreaded) - so I'm trying to work out at the moment how to split the decompression step into chunks.

Here's some ridiculous tiff files:

https://we.tl/ZAH11QapAu

There's a huge photo (LargePhoto.tif), a file representing the largest TIFF images we can currently process (RidiculousTiff.tif) as a benchmark, and the image I'm currently testing with (TypicalImage.tif) which is a fairly typical configuration to the images we'd like to process.

As a reference, with the 'typical' image, our current application would be able to process this in ~2-3s after first uploading, and update our preview in < 100ms after that. Typical upload time for this image would be 15-30s, so the client side application needs to decode and display the TIFF in under this time on a medium spec computer. Photoshop takes ~8s to open the image on my machine.

As another reference, the largest (raster) image files we can actually process are even bigger - jpegs we can currently process files ~ 10x larger in the same amount of time (e.g. 32,000 x 40,000px).

**Comment by photopea** [comment]
Hi, I just made LZW decoding about 5x faster. Update your version.

Note, that you can open your large TIFFs at www.Photopea.com (if WebGL crashes, disable it at More - Use WebGL).

**Comment by manticorp** [comment]
Wow! That's a considerable speed up, I will update my version and give it a try. 😁

It being Christmas might take me a while, but thanks for the updates, help and input!

**Comment by manticorp** [comment]
I'm seeing a good speedup - I don't know how you did it, but it's working nice! Thanks for the update.

Files that were taking 30s to open now are taking about 10s. This is great!

Thanks again.

**Comment by photopea** [comment]
This is weird. A file, that took me 14.5 seconds to load before, takes 2.8 seconds to load now ... I guess it depends on type of data (whether the compressed size is 2x smaller or 20x smaller).
